### DISCUSSION TITLE
10 years of OpenMVG - Sharing some feedback on my experiences

### DISCUSSION THREAD

**Discussion by pmoulon** [openMVG/openMVG#2165] [General]
# Introduction

10 years ago I released the[ first version of OpenMVG](https://github.com/openMVG/openMVG/commit/5c5c221ca8e93a58eb1d1dd83394f75b5f43f48b) Feb 8, 2013 and I thought it would be nice to summarize my learning on 10 years of OpenMVG to celebrate these milestones.

Let’s embark on an introspective journey of my experience adding ’openMVG’; a photogrammetry/3D reconstruction toolkit to the world.

## What is OpenMVG?

OpenMVG was created to extend awareness of the power of 3D reconstruction from images/photogrammetry by developing a C++ framework. I wanted to simplify reproducible research with easy-to-read and accurate implementation of state-of-the-art and "classic" algorithms. OpenMVG is designed to be easy to read, learn, modify and use. Thanks to its strict test-driven development and samples, the library allows building trusted larger systems.

I started to work on OpenMVG a little bit before my PhD, released it during my PhD and then continued to extend it along multiple years working in the industry. Along those years OpenMVG has been used for research and building products.

Today's stats

* 4.6k GitHub Stars
* 1.5k GitHub Forks
* 268 GitHub Watchers
* 15 releases
* 90 Contributors
* +700 academic citations
* Most famous fork/derived products
    * Alicevision/Meshroom
    * Regards3d

I started the project with the following initial motivation, creating an open source project for:

* "Sharing knowledge and state of the art results"
    * Show to the world what the tech can do
    * Share my passion for the field
    * Share knowledge and skills
    * Expectation about collaborations, feedback, and building connections

* A commitment to a high quality engineering standard
    * Create code
    * Setup a build system and continuous integration
    * Add documentation
    * Build a community
    * Enhance reproducibility for multiple view geometry research

# Was it easy… let's see…

First, let's list the difficulties I faced and then reflect on them

* [Build the necessary code and start from scratch](#scratch)
* [Make the entire toolchain open and usable freely](#free)
* [Maintain and contribute to the project in a long term manner to follow the state of the art evolution ](#maintain)
* [Getting contributions](#contrib)

# Building from scratch <a name="scratch"></a>

What does it take to write a SfM library from scratch? => A lot of thinking and effort…

* Basically you have to build an end to end system for 3d reconstruction from images.

Starting from scratch made me face two main challenges:

* Work on the technical complexity of the system being built,
* Work on non-technical/strategic challenges. “The licensing and code usability would be examples of non-technical and strategic considerations respectively. “

## Work on technical complexity

Building a new project from scratch makes you think on how to go from point A to B. You have to identify the data structures, algorithms you need and you quickly realize that you have to code A LOT OF THINGS. For Structure From Motion and Multiple VIew Geometry the High level summary is the following:

* Image IO and processing
* Feature matching
  * Feature detection and description
  * Nearest Neighbor computation
    * Binary and scalar
* Unordered Feature Tracking
  * Link two view matches to N-view matches
* Multiple View Geometry
  * Minimal solvers
    * Relative (Fundamental, Essential, Homography)
    * Absolute (Pose estimation)
  * Motion Averaging
    * Rotation
    * Translation
* Robust Estimation (Ransac)
    * Consensus evaluation
    * Sampling
* Triangulation
  * Geometric vs. algebraic
* Camera model
  * Projective geometry and distortion
* Bundle Adjustment/Nonlinear optimization
* Pipeline
  * Code must be usable as library and also as binary to make people use it
* Build and packaging
  * Provide the ability to make your code compile, run and be delivered to various platform

If you want to tackle all of this alone, it will requires you to learn a lot of new skills and you will have to ensure that each module is working as expected.

=> 3D reconstruction at the end of the day is still a `black box`, we all know the ingredients, but to make it fast, very robust and accurate you need some serious oil.

## Work on non-technical/strategic challenges

Once you identify what you want to build, there are basically two possible paths using existing libraries, or write your own.

* Using existing library leads you to evaluate
    * the complexity of integration
    * the level of trust you can have in the component
    * the licensing consideration and side impact on your project
* Write your own, is challenging but can also be rewarding
    * full ownership of the code but risks of making errors

For OpenMVG the answer was often easy since in 2012/2013 most of the magic requirements list listed in section A was not existing as C++ libraries so I built the necessary part to link them as proof-of-concept and transform them later into cleaner pipelines. To ensure code was doing what it was expected, I used Test Driven Development (TDD). Each OpenMVG modules are implemented with unit tests.

**Take home message** Writing interfaces and algorithms on various topics was a LOT of learning and practice (failures and success are part of the game). You have to look at what exists out there and make decisions (use existing or write your own). In my case, I will be honest, it was challenging, asked for an extreme time commitment, but extremely rewarding at the end.

# An open and permissive codebase <a name="free"></a>

Choosing a license can have a long impact on your project and community. As I already touched on, my goal for OpenMVG was and still is to create knowledge and a framework usable by industry, academic researchers and hobbyists. Upon those considerations I was having some alternative on using MIT, BSD-like, MPL2, …

## Choosing the [MPL2](https://choosealicense.com/licenses/mpl-2.0/) license

The MPL2 license enables people to freely be able to create side projects with the license of their choice but still force contributors to share back the modification/bug fix of the existing code.

MPL2 license is included in each file

* It helps protect header only files

MPL2 allows customization of the copyright header per file

* Each contributor can claim his creation/ownership on a specific file if they want

The main decision points for me was along:

* Protect header files efficiently (openMVG have multiple header file libraries)
* Create a community effect by making people have to share back code/bug fix in the existing code
* Empowering others to use the project without constraints for personal or commercial projects

Note: MPL2 is also the license chosen by [Eigen](https://eigen.tuxfamily.org/index.php?title=Main_Page)

## Side impacts on creating, maintaining a permissive library

Working on creating permissive code can lead to some constraints, it means that all the code and logic you are using is free and not linked to any protection. It means you can use 3rd party projects with contaminating libraries (GPL) and you can’t provide code that implements patented code.

### Code and patent example

Let's take [SIFT](https://en.wikipedia.org//wiki/Scale-invariant_feature_transform) as an example; As most of you know SIFT is one of the best techniques to match keypoint between images. But it was also subject to a patent until not a long time ago. So using it was NO for a permissive library, but a must to have for the research community.

* To keep the project totally permissive, free of patent, we worked on providing a keypoint detector/description patent free alternative by providing our own implementation of [AKAZE](https://github.com/openMVG/openMVG/blob/develop/src/openMVG/features/akaze/AKAZE.hpp).
    * We chose to make our own implementation of AKAZE for 2 reasons
        * not being dependent of OpenCV
        * have code that is much less memory hungry than the original implementation
* On a side note SIFT was still available in OpenMVG => to target the research community
    * As sharing knowledge to the community with a [clear and detailed implementation](https://github.com/openMVG/openMVG/tree/develop/src/openMVG/features/sift)

### Copyleft (protective licensing) example

Being permissive is constraining you on the 3rd party library you can use. I have two particular examples I can cite here.

* In 2010 [SBA ](http://users.ics.forth.gr/~lourakis/sba/)was the only open source Bundle Adjustment library but it was GPL licensed (no go for permissive code). Luckily for me Ceres solver was released with a permissive license in 2012 (perfect timing for this project!).
* When I added [lInfinity](https://github.com/openMVG/openMVG/tree/develop/src/openMVG/linearProgramming/lInfinityCV) solvers (convex or pseudo convex solvers) for computer vision problems, it required [Linear programming solvers](https://github.com/openMVG/openMVG/tree/develop/src/openMVG/linearProgramming). I quickly faced the reality that commercial solutions are much faster and robust/stable than open source solutions. In order to provide something stable for the community I got to effectively test/compare multiple open-source alternatives. In the end sticking to [COIN-OR](https://www.coin-or.org/) was the best option I tested for the speed vs robustness balance.

**Take home message** My willingness to work creating a permissive project providing a full patent free pipeline for 3d reconstruction made me work on finding and implementing alternatives and ensuring they were still reliable compared to the state of the art. It means that I have worked harder to find and implement robust alternative solutions for the community. Despite I understand the need of copyleft and agree with its necessity for some project to survive (i.e CGAL), writing a large permissive license project made me reflect a lot about how the community is doing open reproducible research, publishing work and how patenting can be limiting creativity and accessibility of techniques. As of today, I would encourage computer vision researchers to share more openly and permissively their code and try to integrate their findings in larger frameworks to empower large groups of people.

# Long term maintenance <a name="maintain"></a>

Creating and maintaining an open source project is a lot of personal effort. You basically have to work on every front to setup and maintain each of the following buckets active:

* Features
* Build and continuous integration
* Documentation
* Feedback and project direction

You are basically doing the work of Individual contributor and project manager.

=> In order to maintain a safe direction for the project you need to stick to bold decisions for the project.

## Define project philosophy and direction

Defining a line of actions and focus is very important. It helps you and your community to align to the same goals. So they can decide if they follow along, discuss with you and know on which boat they are going. Here some of the bold decisions I made for openMVG:

* Define simple directions for the projects:
    * "Keep it simple, keep it maintainable"
    * Create easy to read, learn, modify and use
    * Following Test-Driven Development to help build trusted larger systems
* Do not rely on opencv. Opencv was not and is still not a library for multiple view geometry, neither provided interesting feature detection/matching capability at that time (2013). I did not see why rely on this libs only to load/store images pixel data.
* Do not write GUIs and stick to command line binaries to provide more flexibility to users. For beginners Python binaries pipelines launchers were created.

## Risk versus rewards

As a developer and scientist your judgment and skills will evolve and it means that along the development of your project there is opportunity to do better.

* Adding some abstraction and high level abstraction for SFM pipelines is one of the best decisions ever made in OpenMVG (aka the sfm_data container and features/regions abstract data provider). Leveraging this abstraction allowed to simplify a lot of iteration in adding new processing pipelines since a lot of code was shared and quickly updated between different SfM pipelines). I broke the existing APIs but it had a multiplier effect in terms of productivity, code cohesion and usability for coders and users for many years forward!
* At one point I decided to implement algorithms that were not widely having open source implementations to help reproducibility. I learned a lot of new skills and was able to test those algorithms side by side for my research but in the end I did not see a lot of adoption.
  * implement LInfinity CV solvers and a generic wrapper for Linear program solvers.
  * implement native support of spherical panorama for SfM.

## Funding?
* Employer support
    * Luckily being in a PhD program helped a lot at first. I was financed to do research and my partners (academic and industrial) were supportive.
* Working on the same type of feature for work and open source
    * My contributions to OpenMVG were discussed with my employers. Some employers were supportive, others were not really.
* Situation comedy
  * Most of my employers wanted me because of the visibility of my work and knowing I was capable because of the code I wrote and updated over the years. You see the loophole about great work that makes you desired, but some opportunities can push you away from your open source life and project (Intellectual Property and work contract). Just be strong and honest to your employer. If they want you it is for what you are, can do and what you believe in, they must not shape you to someone else for benefits (salary, prestige).

**Take home message** In conclusion, creating and maintaining an open source project requires a significant amount of personal effort. It will require making bold decisions to keep the project running and pleasing to the audience. It is important to have a clear plan for the project and to be open to feedback and support from the community. Your vision defines what is your MVP (Minimum Viable Path/Product), what you think is meeting your quality needs. As in all tasks related to better engineering and quality, the path won’t always be easy, but it is also important to continue to enjoy the process. It is crucial to stay true to oneself and not let the project or pressure shape one into something that is not desired.

# Community building and contributions <a name="contrib"></a>

Without adoption, an open source project does not really exist! To drive adoption, you need to consider building a community, understand their needs, and patiently answer their requests. You will have to be patient and find ways to demonstrate that your work is valuable.

## Building and maintaining a community

Marketing your work is not easy, but if you don’t do it you can’t drive awareness

* Luckily for us, sometimes others are caring about this for you. I.e I got invited to show OpenMVG at CVPR 2015 SFM workshop and it was a great exposure for the project

Caring is making people trustworthy about your project

* One of the reasons why the OpenMVG community exists is because we handle anyone with the same care. You can be a hobbyist or a researcher, the OpenMVG community will take time to onboard you and answer your questions. We want to educate as much as we want people to succeed.

## Getting contributions

In the end it is extremely surprising that many people don't contribute (afraid to commit time, effort, code). People take open source code for granted. They do not realize the efforts set by a group of people to make it available. Having access to something is already a gift that you perhaps never did on your side, so don't judge, acknowledge and respect the work done, share your constructive criticism and add your contribution to the project.

Please establish connections and cultivate, and cherish them. One of the aspects I’m really proud of along with the OpenMVG project is all the human connections it helped me create.

* I must admit that I'm very glad that I have been able to discuss with researcher like Dan Cernea, Chris Sweeney, Venu Madhav Govindu, Noah Snavely, Sameer Agarwal, Ricardo Fabbri, Yasutaka Furukawa, Johannes Schönberger… and many others as they were all instrumental on open source computer vision and multiple view geometry. Their work always inspired me!
* I’m also very glad to have met and worked with amazing people such as Romuald Perrot @rperrot, Romain Janvier @rjanvier, Klemen Istenič @hcjghr, Stéphane Flotron @sflotron, Mitja Puzigaća @mitjap and many others that changed my perspective on how to write good open source code and interfaces.

Sometimes you're lucky! Contributors will come, match the project vision and be instrumental to your future. OpenMVG got lucky and we can count 90 contributors to the project. I’m amazed that multiple contributors came and wrote large and long lasting contributions (it does not happen on every OSS project). What matters for a project is that activity keeps going, whatever it is, usage, feedback and code, they are all helping the project. OpenMVG would not be what it is without all those contributions and contributors!

Here is a piece of advices:

* If you want to start contributing to a project, please do start with just showing that you care. The main point here is that you need to engage, to move yourself out of your comfort zone. Once you start you will like it.
* Soliciting feedback (sometimes with referrals) helps to create more chances at opportunity which may look like luck but is actually luck + initiative.

## Enjoy the journey

Up and down situations will happen so please do celebrate the win that you did and that your community can have. Here some happy surprise that happen for OpenMVG:

* In an attempt to better get to know OpenMVG customers I launched a survey “[2019 Survey](https://github.com/openMVG/openMVG/issues/1659#issuecomment-576446636)”. I’m glad I did it since I learned a lot about how people were using the project and how to better shape the project for the future.
* OpenMVG is actually listed as a skills by many on LinkedIn
* Your community can go above and beyond to use your project:
    * OpenMVG was used to perform mapping and re-localization in all San Francisco ([Collaborative Augmented Reality on Smartphones via Life-long City-scale Maps](https://arxiv.org/abs/2011.05370v1))
    * OpenMVG was used to help injured animals by creating custom animal prosthetics and performing facial reconstruction of century old 3D skulls, see [here](https://arc-team-open-research.blogspot.com/2018/10/openmvgscript.html).
    * OpenMVG is used to perform archaeological surveys [ATOR](https://arc-team-open-research.blogspot.com/2018/10/openmvgscript.html)
    * OpenMVG is still a very good option for mapping [OPEN-SOURCE IMAGE-BASED 3D RECONSTRUCTION PIPELINES: REVIEW, COMPARISON AND EVALUATION ](https://www.int-arch-photogramm-remote-sens-spatial-inf-sci.net/XLII-2-W17/331/2019/isprs-archives-XLII-2-W17-331-2019.pdf)

**Take home message** Any developer should not have fear of exposing his work to others. Being willing to help and try is already a commitment that needs to be celebrated. Anyone should be able to start sharing imperfect work without fear; improve faster due to feedback from people that care; find fans and customers who enjoy the work done. Finally, easier than said, keep in mind that it is important to be open to feedback and consider it as a way to improve rather than a personal attack.

What matters the most along the journey of an open source project is the community factor. If your project is having no traction there is no collaboration, no feedback, no direct future. As developers we often ignore the marketing aspect - how can we showcase our work and make our project more compelling to others to be used and modified.

# Conclusion

Open source software development can bring numerous benefits, including personal and professional growth, increased skills, and potential job and speaking opportunities. To maximize the rewards, it is necessary to approach it with a long-term perspective, actively work on improving the project, and be open to feedback and collaboration. Building a community is crucial for increasing the project's visibility and ensuring its sustainability. Realistic expectations of success should also be considered, as not all projects will achieve widespread popularity. However, the process of creating and maintaining an open source project can be highly rewarding in its own right, beyond just the end result.

## Some additional final thoughts

Computer vision and multiple view geometry algorithms are not yet widely available or easy to use for the general public. As an industry, we still face challenges in creating code that is optimized for various computing architectures in terms of speed, accuracy, power, and memory usage. This is why so many of us are rewriting, adapting code. The goal of being able to simply "pose the problem" and receive an optimal solution for a given computing platform has yet to be achieved.

OpenMVG is my attempt to create a project that allows individuals to learn 3D reconstruction from images. Through working on the project, I have gained valuable knowledge on various fronts and have greatly enjoyed the experience. Sharing this journey with others, discussing and collaborating has been extremely rewarding and helped me being a better developer, scientist and better person.

**Comment by sflotron** [comment]
Thanks @pmoulon it was a pleasure to work with you !

**Comment by luxigo** [comment]
Hey Pierre & Stéphane ! Glad to have believed in both of you ten years ago :-)

**Reply by pmoulon** [reply]
Thx @luxigo it is a bit sad that our shared fun did not last years.. But that's life

**Comment by donlk** [comment]
After so many years, this is still the best open-source sfm tool out there (I've tried them all), mainly due to it's compatibility with other evolved mvs solutions (colmap, openmvs). You've done an incredible job Pierre.

(It's a shame though that the GPU feature extractors did not get finished/implemented, especially now that SIFT is free. There are at least 
3 good CUDA/OpenGL SIFT implementations out there, albeit not in active development).

**Reply by pmoulon** [reply]
GPU as never been my priority. For the main reason that is even harder to have something working on all platform (resonate with my comments on we still as a community did not master how to convert code to be efficient for all type of hardware/platform). But plugging things is really easy in OpenMVG thanks to abstraction that is available at many level.

**Comment by jiabinliang** [comment]
hi Pierre, base on your inspiring global sfm idea, we built a MVS software that can help drone to do many incredible thing that it can not before.  Your work is actually helping people saving lives. Hope one day I can meet you in person and express my gratitude.

**Reply by pmoulon** [reply]
Thank you. Really happy to see that open source research can inspire and lead to product helping people.

**Comment by UkuKert** [comment]
Thank you @pmoulon! I used OpenMVG in my bachelors' thesis and I couldn't have written it without it. Thank you so much for writing this amazing library.

**Comment by Asherchi** [comment]
yours works is great!!!
